src/vision/status_quo/alan_iteratively_regresses.md

```diff
@@ -0,0 +1,90 @@
+
+For the memory allocator issues, Barbara recommnds that Alan try to eliminate the allocation calls and if they cannot be eliminated, then that Alan try tuning the parameters for the global memory allocator, or even switching which global memory allocator he is using. Alan looks at Barbara in despair: his time tweaking GC settings on the Java Virtual Machine taught him that allocator tuning is often a black art.
+
+Barbara suggests that they investigate where the calls to `memcpy` are arising. From the call stacks in `perf report`, Alan and Barbara decide to skim over the source code files for the corresponding functions.
+
+Upon seeing `#[async_trait]` in Alan's source code, Barbara recommends that if performance is a concern, then Alan should avoid `#[async_trait]`. She explains that `#[async_trait]` [transforms][async-trait transform] a trait's async methods into methods that return `Pin<Box<dyn Future>>`, and the overhead that injects that will be hard to diagnose and impossible to remove.
+
+They continue looking at the code itself, essentially guessing at potential sources of where problematic `memcpy`'s may be arising. They identify two potential sources of moves of large datatypes in the code: pushes and pops on vectors of type `Vec<DistriQuery>`, and functions with return types of the form `Result<SuccessCode, DistriErr>`.
+
+Barbara asks how large the `DistriQuery`, `SuccessCode`, and `DistriErr` types are. Alan immediately notes that `DistriQuery` may be large, and they discuss options for avoiding the memory traffic incurred by pushing and popping `DistriQuery`.
+
+For the other two types, Alan responds that the `SuccessCode` is small, and that the error variants are never constructed in his benchmark code. Barbara explains that the size of `Result<T, E>` has to be large enough to hold either variant, and that `memcpy`'ing a result is going to move all of those bytes. Alan investigates and sees that `DistriErr` has variants that embed byte arrays that go up to 50kb in size. Barbara recommends that Alan look into boxing the variants, or the whole `DistriErr` type itself, in order to reduce the cost of moving it around.
+
+Alan uses Barbara's feedback to box some of the data, and this cuts the `memcpy` traffic in the `perf report` to one quarter of what it had been reporting previously.
+
+However, there remains a significant performance delta between the Java version and the Rust version. Alan is not sure his Rust-rewrite attempt is going to get anywhere beyond the prototype stage.
+
+## 🤔 Frequently Asked Questions
+
+*Here are some standard FAQ to get you started. Feel free to add more!*
+
+### **What are the morals of the story?**
```

Comment:
Some other potential morals:
* Lack of certain language features leads to potentially using constructs like `async_trait` which may perform poorly and for which its hard to understand the performance impacts of its usage. 
* In addition to native code tooling understanding C/C++ better than Rust, they also definitely do not have insight into Rust's async model which is even further abstracted from C/C++.

Reply:
thanks, great points! I added them in 0e5819b18932e1b13b91c34fa28acd89608f48cb